Move qr/qc-evp/con and qitot-sub/dep to individual subroutines.

AaronDonahue opened the pull request:
Move the evaporation/condensation of rain/liquid and the deposition/sublimation of ice to individual subroutines.

Change the calculation of ice deposition, ice sublimation, the Bergeron process and rain evaporation to be consistent with the sub-grid variability handled by SHOC and other macrophysics schemes.

This commit is Non-BFB because it removes any super-saturation adjustment that was being calculated inside P3 and removes rain and liquid condensation tendencies as well as liquid evaporation.

AaronDonahue commented:
@PeterCaldwell and @pressel , I've marked this as WIP since I plan on eventually having this PR handle super-saturation consistently with the macrophysics.  This first commit is BFB since it follows in the task that Kyle has been working on and separates the tendencies into their own subroutines.  From here we should be able to focus on the ice and liq separately.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -960,6 +960,8 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
                qrheti,nrheti)
 
+          ! AaronDonahue: The following should be removed or moved to a separate
+          ! subroutine? 
```
I like having a reminder that rime splintering is a thing in the code. It shouldn't be in a separate subroutine since we want the top-level to remind us of the processes one might include. I thought I put a comment in here saying rime splintering is omitted when only 1 ice category is used to compensate for the fact that ice is being forced to fit a single distribution? I guess I'm saying that we should leave the comment as is.

AaronDonahue replied:
Does that extend to the calculation of aaa and xx?  In the original code (which will be changed) both aaa and xx are used for the rain/liq calculations and for the ice calculations so I decided to keep them separate from the individual subroutines.

PeterCaldwell replied:
I'm confused why you bring up aaa and xx. I was just saying we should keep the ice multiplication header comment and my subsequent comment intact. Are you envisioning deleting a much larger swath of code? If so, can you somehow indicate the start and end points of the code you want to remove? Or we could talk about it in person...

AaronDonahue replied:
i only brought it up because the full effort is to compartmentalize the code in such a way that we have a set of individual subroutines and this is a big chunk of code. We can’t remove it, i’m just wondering if we want to make a separate subroutine for calculating xx and aaa to keep this part of the code clean.

AaronDonahue replied:
I've moved the calculation of xx and aaa to it's own routine.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -972,38 +974,10 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
+          epsr,epsc)  
           !PMC moved oabi outside t<273.15 loop. oabi is only *used* where t<273.15, but
           !t could be updated before oabi is used, resulting in unititialized use. Note
           !that t is not currently being updated before oabi is used, so this is just future-proofing.
```
Can my oabi comment be deleted yet?

AaronDonahue replied:
I think so.  This whole calculation may not be needed when we switch to handling super-saturation using the MG code.  But now that it is brought up, do we want to use the calculation of ab directly from P3 or calculate ab on the fly using the MG routine from the utils?

PeterCaldwell replied:
My understanding is that ab needs to be recalculated every time T changes. I hope/expect it to be computed identically in P3 versus MG. If not, we should look at how it's calculated but probably use the P3 version.

AaronDonahue replied:
For the most part this is true.  As we noted in our email to Hugh, we are not updating the temperature to reflect any impact from clipping at the beginning of P3.  After we hear back from him about this I suspect we will fix this potential bug and ab and abi will reflect the most up-to-date temperature.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1054,61 +1028,19 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
-          endif
+          ! AaronDonahue: The below subroutine should be moved to within
+          ! evaporate_sublimate_precip, but doing so leads to a non-BFB
+          ! solution.
```
I don't understand why putting the _limiter stuff at the end of evaporate_sublimate_precip isn't bfb, but I'm sure you've confirmed this is the case. You should, of course,  put _limiter stuff into evaporate_sublimate_precip as a future PR even though it's not bfb.

AaronDonahue replied:
I'm not sure why either, and it has driven me a bit batty :-)

I agree that the limiter stuff should go directly into the evaporate_sublimate_precip subroutine and since the PR itself will not be BFB it shouldn't be a problem to do it right.  I brought it up here hoping that someone else might ameliorate my confusion about why its not BFB when I do it the correct way.

ambrad replied:
What if you compile with CMAKE_BUILD_TYPE=Debug or SCREAM_FPMODEL=strict? BFB-ness, particularly with CMAKE_BUILD_TYPE=Release but SCREAM_FPMODEL=strict, would then imply it's a vectorization difference.

pressel replied:
@AaronDonahue I'll experiment with this in P3 standalone a bit tomorrow. I think we should be able to quickly do a binary search to find which lines are breaking BFB. It will also be useful to see if P3 standalone compiled with gfortan/g++ has similar behavior.

AaronDonahue replied:
@ambrad , running with DEBUG compile fixed the BFB issue, so it may be a vectorization difference.

AaronDonahue replied:
since DEBUG compile is still BFB I have resolved this comment and consolidated the two routines into one.

AaronDonahue commented:
@pressel , in this PR I have now changed one of your earlier routines for ice self collection.  Go ahead and take a look to see if you agree with this change.

pressel commented:
@AaronDonahue I agree with your change to ice self collection.

AaronDonahue commented:
@PeterCaldwell , this newest commit changes how ice deposition and sublimation is handled.  There are a few comments included in the ice_deposition_sublimation subroutine and one more in the p3_main code that we may want to consider before fully switching to this approach for these tendencies.  When you get a chance could you take a look?

AaronDonahue commented:
Also, for the `qiberg` variable, I see that in MG this tendency is multiplied by `micro_mg_berg_eff_factor` which turns out to be 0.1 by default.  Is this something we should be doing too?

PeterCaldwell commented:
I think micro_mg_berg_eff_factor doesn't make physical sense, so I'd rather not include it. I'll look at the other comments now.

PeterCaldwell commented:
line 1229-1230: this is how supersaturation was limited to not remove more mass than available by saturation adjustment. It assumes there's no sub-grid moisture variability, so we need to get rid of it because it is inconsistent with SHOC. I don't think we need any checks here.

AaronDonahue commented:
Ok cool, then I won't bother with adding in a berg_eff_factor.  As for lines 1229-1230, that was my feeling, I'll go ahead and remove these lines.

keeping up with my questions looking at MG and comparing what we do, I also noticed that in MG, berg, vap_dep (qidep) and ice_sublim (qisub) are done with cell-avg quantities.  maybe because they are dependent on the saturation mixing ratios.  Should we treat them the same way and not multiply by the cloud fraction later?

PeterCaldwell commented:
Answers for lines 3274-3377:
1.	I’m flummoxed why dt would show up in qidep. Have you found the equation that’s being solved in the original paper? That might explain things. The new formulation looks like I would expect.
2.	The supi very small thing is just saying if you’re not supersaturated with respect to ice, get rid of all ice in a single timestep. This is a numerical superstition which is rampant in this code for reasons I don’t understand. I don’t think we need to enforce it.
3.	Yeah, I think the code is depending on initialization to zero. It is probably better practice to set those values to zero here as well.
4.	Ah - I think I understand the rationale for P3-WRF using qv - dumqvi while we use qvs (which I assume is liquid saturation mixing ratio): If you didn’t have subgrid variability, the only way for Bergeron to operate would be if you had liquid cloud, which would mean that cell-ave qv would naturally be larger than qvs. Only at the coarse resolutions we’re working with now and without the assumption that P3 does condensation/evaporation would qv be significantly different than qvs. I think changing qvs to qv is essentially fixing what was a totally inappropriate choice in the old version. Can you confirm that qv << qvs in the cases where you’re seeing big changes in qidep and that our switch to qvs is bringing qidep from a negligible value to the same order of magnitude as other terms?

PeterCaldwell commented:
> keeping up with my questions looking at MG and comparing what we do, I also noticed that in MG, berg, vap_dep (qidep) and ice_sublim (qisub) are done with cell-avg quantities. maybe because they are dependent on the saturation mixing ratios. Should we treat them the same way and not multiply by the cloud fraction later?

Cell-average seems weird to me. vap_dep can only occur in the portion of the cell with cloud ice to deposit onto. Doubly so for berg. I'll think about it some more, but right now I'd vote for doing these things in-cloud.

This also brings up the question of how we're handling precip_area_frac for P3. This question gets weird because we don't have a separate category for snow, so do we assume there's cloud wherever we have precipitating ice?

AaronDonahue commented:
**Response to the cell-avg comment:** I misspoke in my comment in the code.  Upon further investigation the interface handles everything as cell-avg but where the tendencies are calculated  it looks like it uses incloud values, it just calculates the incloud amount inside the function, this is then used to find the relaxation timescales and after that we go back to treating things as cell-averaged.

AaronDonahue commented:
**Response to qv vs qvs comment:** I think I crafted my comment in a confusing way, my apologies.  MG uses qv and P3-WRF uses qvs.  I ran a ne30 simulation and recorded all the info needed to construct qidep, qisub and berg using both approaches.  What I found was:

If we consider that qidep and qisub are two sides of the same coin, in overall magnitude the new approach taken from MG results in tendencies that are on average an order of magnitude greater than the P3-WRF version, and in terms of extremes the largest qisub in the new approach is 4 orders of magnitude greater than the largest qisub in the P3-WRF version.

I think the most interesting thing which we can see from the code itself is that the qidep term in P3-WRF is equivalent to the qiberg term taken from MG.  I guess the question is, in P3-WRF do they consider vapor deposition to simply be an application of the Bergeron process?  Does this even make sense?

AaronDonahue commented:
@PeterCaldwell and @pressel , I've now added the changes to evaporation and sublimation of liquid cloud that should complete the transition of P3 away from handling super-saturation.  I've left the old code in as reference and for testing, but if you could take a look at what I've changed in this latest commit and provide comments that would be great.  When we are all satisfied that these changes are correct I'll go ahead and remove the old code and we can merge this PR.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -3446,6 +3443,39 @@ subroutine evaporate_sublimate_precip(qc_incld,qr_incld,nr_incld,t,pres, &
+      qrevp = qrevp/rcldm
+   end if ! rcld>cld
+   if (qr_incld(i,k).gt.qsmall)  nrevp = qrevp*(nr_incld(i,k)/qr_incld(i,k))
+   ! No condensation, or evaporation of liquid
```
this comment might be better phrased as "We are assuming macrophysics handles condensation/evaporation of qc and that there is no condensation of rain. Thus setting below values to zero"

AaronDonahue replied:
I guess when we switch entirely should we simply remove qccon, qcevp and qrcon from P3-SCREAM completely?

PeterCaldwell replied:
I hadn't thought of removing them entirely, but yes we might as well.

AaronDonahue replied:
I'm happy to delete them all since carrying around and applying a bunch of variables set to 0 is kind of sloppy.  Once we do this it may be difficult for a P3-WRF person looking at the code to understand what happened to the evaporation and condensation variables they would expect.  I propose removing them entirely but having a comment in the microphysics tendencies declarations (or in the evaporation_sublimation... subroutine) that states that these processes have been removed because (see your comment above).  Does that sounds reasonable?

PeterCaldwell replied:
Yeah, I think deleting them and leaving a comment is the way to go. Thanks for doing that.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -3367,27 +3359,24 @@ subroutine ice_deposition_sublimation(qitot_incld, nitot_incld, t, &
-subroutine evaporate_sublimate_precip(qc_incld,qr_incld,nr_incld,t,pres, &
-aaa,xx,dt,ab,epsr,epsc,qv,sup,ssat,xxlv, &
+subroutine evaporate_sublimate_precip(qc_incld,qr_incld,nr_incld,qitot_incld,t,pres, &
+lcldm,rcldm,qvs,aaa,xx,dt,ab,epsr,epsc,qv,sup,ssat,xxlv, &
```
Kind of an obvious point, but once you delete orig_version, you can delete a lot of variables from this interface.

AaronDonahue replied:
Absolutely.  But always good to note.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1220,26 +1220,13 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
           ! cloud
           sinks   = (qcaut+qcacc+qccol+qcevp+qcheti+qcshd+qiberg)*dt
           sources = qc(i,k) + (qccon+qcnuc)*dt
+          ratio = 0.0 ! Initialize ratio in order to only limit qidep and qisub when needed.  See below.
```
I still don't see how ratio = 0.0 is ever used. Won't it be set 100% of the time in the if/else below? Also, shouldn't it be _rtype if we did keep it?

AaronDonahue replied:
Ahh yes good point, this is actually an incorrect fix.  The idea was that if there is qc but no sinks we don't want the next if statement to remove all qidep and qisub, in other words we want ratio to be zero.  But you are right, as it is currently set up ratio will always be set in sinks vs. sources if statement.  I think the correct fix is to add an elseif for the special case where qidep and qisub shouldn't be limited.  I'll make the change.

AaronDonahue commented:
@PeterCaldwell , thanks for looking over the most recent addition.  I've also gone through Hugh's response to our questions about epsr and epsi and am satisfied that we are doing this correctly.  Based on that I'm ready to remove the [WIP] designation on this PR and with you and @pressel 's blessing go ahead and merge.

AaronDonahue commented:
@ambrad , do you have any comments on this PR?  Or should I merge?

ambrad commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -177,26 +177,36 @@ SUBROUTINE p3_init_a(lookup_file_dir,version_p3)
 
   subroutine p3_get_tables(mu_r_user, revap_user, vn_user, vm_user)
     ! This can be called instead of p3_init_b.
+    implicit none
```
Not that it matters, but `implicit none` at module level covers all contained subroutines.

ambrad commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -177,26 +177,36 @@ SUBROUTINE p3_init_a(lookup_file_dir,version_p3)
     vn_table(:,:) = vn_user(:,:)
     vm_table(:,:) = vm_user(:,:)
+
+   return
```
I don't mind these, but fwiw, it's nice to reserve `return` for non-default return cases. `end` at the subroutine level has an implicit `return` in Fortran.

AaronDonahue commented:
Excellent, I'll go ahead and merge.  But I'll keep in mind your points about "implicit none" and "return" for future PR's.  Thanks for reviewing.
